examples/hmm_enum.py

```diff
@@ -0,0 +1,279 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Enumerate Hidden Markov Model
+=============================
+
+This example is ported from [1], which shows how to marginalize out
+discrete model variables in Pyro.
+
+This combines MCMC with a variable elimination algorithm, where we
+use enumeration to exactly marginalize out some variables from the
+joint density.
+
+To marginalize out discrete variables ``x``:
+
+1. Verify that the variable dependency structure in your model
+    admits tractable inference, i.e. the dependency graph among
+    enumerated variables should have narrow treewidth.
+2. Ensure your model can handle broadcasting of the sample values
+    of those variables
+
+Note that difference from [1], which uses Python loop, here we use
+:func:`~numpryo.contrib.control_flow.scan` to reduce compilation
+times (only one step needs to be compiled) of the model. Under the
+hood, `scan` stacks all the priors' parameters and values into
+an additional time dimension. This allows us computing the joint
+density in parallel. In addition, the stacked form allows us
+to use the parallel-scan algorithm in [2], which reduces parallel
+complexity from O(length) to O(log(length)).
+
+**References:**
+
+    1. *Pyro's Hidden Markov Model example*,
+       (https://pyro.ai/examples/hmm.html)
+    2. *Temporal Parallelization of Bayesian Smoothers*,
+       Simo Sarkka, Angel F. Garcia-Fernandez
+       (https://arxiv.org/abs/1905.13002)
+"""
+
+import argparse
+import logging
+import pickle
+import time
+
+from jax import random
+import jax.numpy as jnp
+
+import numpyro
+from numpyro.contrib.control_flow import scan
+from numpyro.contrib.indexing import Vindex
+import numpyro.distributions as dist
+from numpyro.handlers import mask
+from numpyro.infer import HMC, MCMC, NUTS
+
+
+# Let's start with a simple Hidden Markov Model.
+#
+#     x[t-1] --> x[t] --> x[t+1]
+#        |        |         |
+#        V        V         V
+#     y[t-1]     y[t]     y[t+1]
+#
+# This model includes a plate for the data_dim = 44 keys on the piano. This
+# model has two "style" parameters probs_x and probs_y that we'll draw from a
+# prior. The latent state is x, and the observed state is y.
+def model_1(sequences, lengths, args, include_prior=True):
+    num_sequences, max_length, data_dim = sequences.shape
+    with mask(mask_array=include_prior):
+        probs_x = numpyro.sample("probs_x",
+                                 dist.Dirichlet(0.9 * jnp.eye(args.hidden_dim) + 0.1)
+                                     .to_event(1))
+        probs_y = numpyro.sample("probs_y",
+                                 dist.Beta(0.1, 0.9)
+                                     .expand([args.hidden_dim, data_dim])
+                                     .to_event(2))
+
+    def transition_fn(carry, y):
+        x_prev, t = carry
+        with numpyro.plate("sequences", num_sequences, dim=-2):
```

Comment:
This is a nit for these models, but these `sequences` plates should really live outside of `transition_fn` to avoid tempting users to potentially break enumeration by enumerating variables inside plates across timesteps.  Were there issues supporting that? You may need to do the same trick with `plate` that you did with `trace` and `substitute`, where you push the `plate` handler inside of the `scan` wrapper, but that doesn't seem so bad.

Reply:
Do you mean to support something like
```
with plate(...):
    scan(...)
```
? If so, I think that it is tricky to put time dimensions at the right position (it seems to be doable but might require non-enum users write correct plate codes). Could you give an example on how things could go wrong so I can add to the docs instead?

Reply:
All of the [restrictions from Pyro's enumeration tutorial](http://pyro.ai/examples/enumeration.html#Dependencies-among-plates) still apply.  A simple failure case would be having an enumerated variable outside of a plate depend on a variable enumerated inside that plate at the previous timestep (`x_t ~ p( . | x_t-1, y_i_t-1)`, `y_t_i ~ p( . | x_t)`).

Reply:
Thanks, I didn't know about those restrictions. I'll add a link to it and learn more about them.